### PR TITLE
feat(lockservice): opt-in PostgreSQL session (advisory) lock service

### PR DESCRIPTION
**Pull request by ymajoros**
StandardLockService keeps the changelog lock as a committed LOCKED=1 row in DATABASECHANGELOGLOCK. If the process dies mid-update (OOM, pod eviction, a startup probe pulling the plug, plain old kill -9), that row stays locked and every run after it bails out with "Could not acquire change log lock" until someone runs releaseLocks by hand. Not fun at 2am, and not really workable in a rolling container deploy.

Session-level locks fix this for free. The lock lives on the DB connection, so the server drops it the moment the session dies. A killed process can't leave anyone locked out.

This adds that for Postgres, fully opt-in:

- SessionLockService: an abstract base on the LockService SPI
- PostgreSQLSessionLockService: pg_try_advisory_lock / pg_advisory_unlock, listLocks reads pg_locks joined to pg_stat_activity
- a new liquibase.useSessionLock property, default false. It only kicks in on Postgres 9.1+ when the flag is on, and sits above StandardLockService by priority, so nothing changes unless you ask for it

Same trick exists on the other engines (MySQL GET_LOCK, SQL Server sp_getapplock, Oracle DBMS_LOCK) and can ride on the same base in later PRs.

A bit of background: the old blagerweij/liquibase-sessionlock extension did exactly this, but it's 4.x only and the 5.0 LockService SPI change breaks it, so 5.x users are kinda locked out (ok, last one).

Tests: PostgreSQLSessionLockServiceTest covers the gating (off by default, on only for Postgres 9.1+ with the flag, ignored on other DBs, outranks the standard service). Glad to add a Testcontainers test under liquibase-integration-tests that actually kills a session and watches the lock disappear, if you want it in here.

Against #1453. I went with opt-in flag plus priority to keep the default untouched. If you'd rather have it as a separate extension or selected some other way, happy to rework. The Postgres impl is already running in prod on our side.

### CODE REVIEW

File: liquibase-standard/src/main/java/liquibase/lockservice/SessionLockService.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    public void releaseLock() throws LockException {
+        if (!hasChangeLogLock) {
+            return;
+        }
+        try {
+            releaseLock(getConnection());
+            Scope.getCurrentScope().getLog(getClass()).info("Successfully released change log lock");
+        } finally {
+            hasChangeLogLock = false;
+        }
```
[the remark this answers is not in the document]
Comment (ymajoros) [reply liquibase-standard/src/main/java/liquibase/lockservice/SessionLockService.java]: Good catch. 5f31e97 clears hasChangeLogLock only after the unlock actually succeeds, in both releaseLock and forceReleaseLock, so a failed pg_advisory_unlock no longer desyncs us from the session lock. Same commit guards the null lock-granted in describeCurrentLock.

File: liquibase-standard/src/main/java/liquibase/lockservice/SessionLockService.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    private String describeCurrentLock() throws LockException {
+        DatabaseChangeLogLock[] locks = listLocks();
+        if (locks.length == 0) {
+            return "UNKNOWN";
+        }
+        DatabaseChangeLogLock currentLock = locks[0];
+        String grantedAt = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT).format(currentLock.getLockGranted());
+        return currentLock.getLockedBy() + " since " + grantedAt;
```
[the remark this answers is not in the document]
Comment (ymajoros) [reply liquibase-standard/src/main/java/liquibase/lockservice/SessionLockService.java:247]: Fixed in 5f31e97. describeCurrentLock now returns just the lockedBy string when lock-granted is null, instead of formatting a null and blowing up on the timeout path.

**Comment by ymajoros** [comment]
Pushed a small follow-up (a11bb78): a false return from pg_advisory_unlock is now treated as already-released (log + clear the held flag) rather than thrown, so a reused service can't report a lock it no longer owns. Genuine SQL errors still propagate.

**Comment by ymajoros** [comment]
Thanks, addressed the out-of-diff findings in 2facc3b:

- waitForLock no longer hot-spins after an interrupt. sleepRecheckInterval now throws a LockException on InterruptedException (interrupt flag preserved), so the loop aborts promptly instead of spinning on lock polls until the timeout.
- Added the two clarifying comments: why pg_locks objsubid = 2 (the two-int4 advisory key form), and that backend_start is only an upper bound on lock acquisition time.

I skipped the suggestion to add unit tests for the unlock-failure / null lock-granted paths: those need a live (or heavily mocked) JDBC session and the current tests cover the selection/gating logic, so it felt like more mock than signal. Happy to add them if you'd prefer.

**Comment by ymajoros** [comment]
@filipelautert @wwillard7800 could you review this? Thanks.

**Review by wwillard7800 — COMMENTED** [review]
# Code Review — opt-in PostgreSQL session (advisory) lock service

**Verdict:** Solid direction, well-scoped, and the comments in the new code are very good. I'd recommend changes before merge — there's one real correctness issue, two smaller correctness bugs, an architectural concern, and a test-coverage gap the author already offered to close.

## Strengths

- **Opt-in flag + priority gating** keeps existing behaviour untouched. `supports()` correctly requires `liquibase.useSessionLock` + `instanceof PostgresDatabase` + PG ≥ 9.1, and `PRIORITY_DATABASE (5) > PRIORITY_DEFAULT (1)` makes `LockServiceFactory` pick it once enabled. Confirmed against `LockServiceFactory.getLockService` (`liquibase-standard/src/main/java/liquibase/lockservice/LockServiceFactory.java:53`).
- **Reuses the existing SPI.** No factory or core-engine plumbing changes — clean.
- **Comments are load-bearing in the best way** — the rationale on hash stability, the `backend_start` upper-bound caveat, why the interrupt is re-raised, and why `releaseLock` only clears `hasChangeLogLock` after a successful unlock are all very useful.
- **Test-framework choice (JUnit 4 + Mockito + AssertJ)** matches existing `StandardLockServiceTest`/`LockServiceFactoryTest`. Good.

## Correctness issues — please fix before merge

**1. Advisory-lock key is keyed on the wrong schema.** `PostgreSQLSessionLockService.getChangeLogLockId()`:

```java
return new int[]{
    database.getDatabaseChangeLogLockTableName().hashCode(),
    defaultSchemaName.hashCode()
};
```

`StandardLockService` creates `DATABASECHANGELOGLOCK` in `liquibaseSchemaName` (`StandardLockService.java:127-131`), which is what defines the lock's scope. `defaultSchemaName` is what `current_schema()` returns for the current connection — the two only coincide when `liquibase.liquibaseSchemaName` is unset (it then falls back to `defaultSchemaName` via `AbstractJdbcDatabase.getLiquibaseSchemaName()` line 715). When a deployment explicitly sets `liquibase.liquibaseSchemaName` (a common, recommended practice — e.g. dedicate `migrations` to Liquibase), the lock's scope and the lock table's location diverge:

- **Over-serialization (operational surprise):** Two apps with `liquibase.liquibaseSchemaName=migrations_a` vs `migrations_b`, both running against the default `public` user schema, compute the same advisory-lock key (`("DATABASECHANGELOGLOCK".hashCode(), "public".hashCode())`). They have independent lock tables but will queue behind each other on the advisory lock. StandardLockService does not do this.
- **Silent loss of mutual exclusion (the bad one):** Two sessions sharing `liquibase.liquibaseSchemaName=migrations` but running as DB users whose `current_schema()` differs (per-user schema deployment, or explicit `SET search_path`). They write to the same `migrations.DATABASECHANGELOGLOCK` but compute different advisory-lock keys → no mutual exclusion. By construction this can only happen when `liquibase.liquibaseSchemaName` is explicitly set, so it's narrower than the over-serialization case — but it's a genuine regression vs StandardLockService for that deployment shape.

Fix: key on `getLiquibaseSchemaName()` (and arguably mix `getLiquibaseCatalogName()` into the upper int4 too). Match the scoping StandardLockService already uses.

**2. `listLocks()` NPEs when `backend_start` is NULL.** `DatabaseChangeLogLock(int,Date,String)` does `this.lockGranted = new Date(lockGranted.getTime())` — no null guard. The PR LEFT-JOINs `pg_locks` to `pg_stat_activity`, so the activity-side columns can be null (system-process pids, races with backend termination, restricted visibility for non-superusers). The unreachable `if (lockGranted == null)` branch in `SessionLockService.describeCurrentLock()` is a tell — someone considered this but the upstream constructor doesn't allow null. Coerce a null `backend_start` to `new Date(0)` (or `Date.from(Instant.EPOCH)`) and adjust the comment, or skip the row.

**3. `liquibase.changeLogLockWaitTimeInMinutes` / `liquibase.changeLogLockPollRate` are ignored.** `SessionLockService` hardcodes `changeLogLockWaitTimeMinutes = 5` and `changeLogLockRecheckTimeSeconds = 10`. The setters only fire when callers explicitly invoke them; the global configs (`GlobalConfiguration.CHANGELOGLOCK_WAIT_TIME`, `GlobalConfiguration.CHANGELOGLOCK_POLL_RATE`) are never consulted. Compare `StandardLockService.getChangeLogLockWaitTime()` (line 80) and `getChangeLogLockRecheckTime()` (line 92) — both use the explicit-setter-wins-else-config-fallback pattern. Use the same approach.

## Architectural concerns

**4. No transaction housekeeping around the advisory-lock SELECT.** Liquibase runs with `autoCommit=false` by default. `acquireLock(Connection)` issues `SELECT pg_try_advisory_lock(?, ?)` and never commits or rolls back. The session-level advisory lock survives transaction boundaries, so the lock itself is fine — but the implicit transaction holds catalog locks and starts the migration's first real DML mid-transaction. `StandardLockService.acquireLock` brackets its work with `database.rollback()` before and in `finally` for exactly this reason. Either match that pattern or document explicitly that the lock service deliberately does not touch tx state.

**5. Test coverage is gating-only.** The unit test verifies "the right LockService is picked," not "the LockService does the right thing." There is no coverage of:
   - `acquireLock(Connection)` issuing the expected SQL with the expected args
   - `releaseLock` honouring the "not held" return code
   - `listLocks` producing a sane `DatabaseChangeLogLock` (including the null-`backend_start` case from #2)
   - the actual session-level guarantee — connection drop → lock released

The author already offered to add a Testcontainers integration test under `liquibase-integration-tests` that kills a session and watches the lock disappear. Yes please — that's the only test that exercises the entire reason this PR exists.

## Minor

- **`describeLockHolder`** can emit `"host.example (null)"` if `state` is null in `pg_stat_activity`. Trivial null check.
- **Dead null branch in `describeCurrentLock`** (see #2) — once #2 is fixed, simplify or remove the branch.
- **`listLocks` reports any advisory lock with this `classid`/`objid`/`objsubid=2` from any session in the cluster, not just our own.** Benign because keys differ across deployments, but worth a one-line comment that this is a cluster-wide view.
- **Hash collisions:** two `int4` hashes are fine in practice but undocumented. Worth a comment that `(table_name_hash, schema_hash)` collisions are theoretically possible but practically negligible for any single deployment.
- **Property naming:** `liquibase.useSessionLock` is fine. A small case for `liquibase.lockService.useSessionLock` since this is a lock-service selector — bikeshed-tier, name it however you like.

## Suggested next steps

1. Fix #1, #2, #3 (correctness).
2. Either fix or document #4 (transaction handling).
3. Add the Testcontainers integration test you offered.
4. Tighten the minor items.

Happy to take another pass once those are in. The feature itself is the right answer to issue #1453, and matches one of the options listed there.

---

_(Edited to fix code-formatting that didn't render in the original post.)_

**Comment by ymajoros** [comment]
Thanks for the thorough review, @wwillard7800. All addressed, pushed in three commits:

1. **Lock key scope.** Now keyed on `getLiquibaseSchemaName()` (with the catalog folded into the same int4), not `getDefaultSchemaName()`, so the advisory key lines up with where the lock table actually lives. (1c3b470)
2. **Null `backend_start`.** `listLocks()` coerces a null session start to the epoch instead of NPEing in the `DatabaseChangeLogLock` constructor. (1c3b470)
3. **Wait/poll config.** Both now fall back to `CHANGELOGLOCK_WAIT_TIME` / `CHANGELOGLOCK_POLL_RATE`, explicit setter winning, same pattern as `StandardLockService`. (f4c69ca)
4. **Transaction housekeeping.** `acquireLock`/`releaseLock` now roll back to end the implicit transaction, matching `StandardLockService`. The session lock is unaffected by the rollback, so it survives. (f4c69ca)
5. **Tests.** The unit test now covers key scoping, the already-released unlock path, the null `backend_start`, and the config fallback. Added a Testcontainers integration test that kills the holder's backend with `pg_terminate_backend` and watches a second session pick the lock up, with no manual `releaseLocks`. I terminate rather than close the session on purpose: a graceful close lets the driver tidy up and would not prove the lock outlives a crash. (7cf4a17)

Minors: guarded a null `state` in `describeLockHolder`, dropped the now-dead null branch in `describeCurrentLock`, and added comments on the cluster-wide view and the hash-collision reasoning. Left the property as `liquibase.useSessionLock`.

Thanks again for catching the schema-key issue before it locked anyone out. Ready for another pass whenever you are.

**Comment by ymajoros** [comment]
Cleaned up two connection-leak nits CodeRabbit flagged on the IT (d11a08b): both sessions now open inside the `try`, and a raw connection is closed if wrapping it in a `Database` fails. Used a success-flag + `finally` instead of catch-and-rethrow, and narrowed the quiet-close to `SQLException` rather than swallowing everything. IT still green. Nothing here it can leak now.

**Review by filipelautert — COMMENTED** [review]
Hello @ymajoros !
Ran the lock tests on this branch and after a local merge with main (the new LockService SPI throws change), all green. wwillard7800's five items are all in.
Two things:
CockroachDatabase extends PostgresDatabase, so with the flag on supports() picks it up too and Cockroach's advisory lock support is partial at best. We now have capability methods on AbstractPostgresDatabase for exactly this (supportsEnumTypeSnapshot etc), so I would add a supportsAdvisoryLocks() there defaulting true, override it to false in CockroachDatabase and check it in supports().
forceReleaseLock can't release a lock held by another live session (pg_advisory_unlock only unlocks your own), so releaseLocks would report success while the holder keeps the lock. Worth a warning log or javadoc note.

**Comment by nightknight77** [comment]
Hello, sorry for jumping in randomly, but thanks for working on this PR! This looks like it could address the “process died / DB session disappeared” stale-lock case really well.

One thing I’m wondering about is pooled JDBC connections. We have a special app-server setup where Liquibase may run on a pooled connection (Quarkus Agroal in our case), so closing the Java `Connection` only returns it to the pool and does not necessarily close the physical PostgreSQL backend session. Since session-level advisory locks survive transaction boundaries and are released only by `pg_advisory_unlock` or session termination, a failed release path could theoretically leave an advisory lock held by an idle pooled connection. Is this understanding correct? 

We observed cases in which a long-lived pod caused a stale lock that neither connection loss nor eviction could explain.

Does this implementation have any protection against returning a connection to the pool while still holding the advisory lock?

**Comment by ymajoros** [comment]
@filipelautert thanks, both items are in, along with a CI fix and one more thing that fell out of them.

**1. CockroachDB** (c744575). Added `supportsAdvisoryLocks()` to `AbstractPostgresDatabase` alongside the other capability methods, overridden to false in `CockroachDatabase`, and `supports()` now checks it, so with the flag on Cockroach keeps `StandardLockService`. `EnterpriseDBDatabase` inherits true.

One deliberate deviation: I defaulted it to true on the abstract as you suggested, rather than false-with-`PostgresDatabase`-opting-up like the four snapshot hooks next to it. Advisory locks are core PostgreSQL since 8.2, so opting down is the exception here rather than the rule. Happy to flip it if you would rather keep the file uniform.

**2. forceReleaseLock** (9dbd85a). Documented the limitation and made it non-silent: after the unlock it re-reads `listLocks()` and warns naming the holder if the change log is still locked.

I kept it a warning rather than an exception. Running `releaseLocks` defensively before an update is a common scripted pattern, and throwing would fail it exactly when nothing is wrong, ie. when a concurrent deploy is legitimately holding its lock. The downside is that `ReleaseLocksCommandStep` still prints its success message afterwards, so the output reads as a warning followed by "Successfully released...". I can either throw instead, or make that message reflect what actually happened, if you prefer one of those. The second touches shared command code that `StandardLockService` also runs through, which is why I did not just do it.

**3. CI was red, and it was mine** (d41000e), unrelated to your review. `StandardServiceLocatorTest` compares each service-loader file against the alphabetically sorted implementation list, and I had appended `PostgreSQLSessionLockService` at the end of `META-INF/services/liquibase.lockservice.LockService` instead of sorted in. That was failing every test job.

**4. A real one that came out of looking at point 2** (345e0be). `releaseLock()` is only reached when we believe we hold the lock, so `pg_advisory_unlock` returning false there means the session lock was lost before we released it: a backend dropped and re-established behind a connection wrapper, or a `DISCARD ALL` on the session. The change log was then unguarded for part of the run that just finished and another process could have migrated concurrently, which is the exact failure this service exists to prevent. That was a warning, and it now throws. The abstract `releaseLock(Connection)` reports whether this session held the lock, so `releaseLock()` can treat false as a lost lock while `forceReleaseLock()` still accepts it as normal.

Worth knowing that this does not actually fail an `update`: `AbstractUpdateCommandStep`, `AbstractFutureRollbackCommandStep` and `LockServiceCommandStep` all catch `LockException` and log it severe, and only `DropAllCommandStep` lets it propagate. Making it genuinely fail the run means changing that shared catch, which changes `StandardLockService` behaviour for every user, so I left it out of a strictly opt-in PR. Tell me if you want it here or as a follow-up.

Also rebased onto current main, it was 57 commits behind. Full `liquibase-standard` suite green locally on JDK 21 (6498 tests), and `liquibase-integration-tests` compiles against the changed `releaseLock(Connection)` signature, which is new in this PR so nothing outside it depends on the old one.

@wwillard7800 you reviewed the earlier round and this touches the release paths you looked at, so flagging it for you too.

**Comment by ymajoros** [comment]
@nightknight77 no problem, and yes, your understanding is right on both counts.

The lock lives on the physical backend session. Closing a pooled `Connection` only returns it to the pool and leaves that backend alive, so the "the server releases it for you" property does not apply to a pooled connection that outlives the run. There the lock is released only by the explicit `pg_advisory_unlock` in `releaseLock`, exactly as you describe.

What the implementation does have:

- Liquibase releases the lock in the command step cleanup, and if that throws, `LockServiceFactory.resetAll()` runs immediately after and `SessionLockService.reset()` retries the unlock on the same connection. The held flag is only cleared on success, so the retry does fire. A single failed release is therefore retried before the connection goes anywhere.
- Since your comment, a `false` return from `pg_advisory_unlock` on the normal release path is no longer logged away. Reaching that path means we believed we held the lock, so false means the session lock was lost before we released it, which is precisely the pooled-connection reconnect case you are describing. It now throws instead of warning, because it also means the change log was unguarded for part of the run and another process could have migrated concurrently.

What it does not have: any way to stop the pool handing that connection back out while the lock is still held. Liquibase only sees the wrapper `Connection` and cannot tell Agroal to evict the physical backend. So if both the release and the retry fail, the lock survives on an idle pooled connection until the pool discards it or the process exits.

Given that, I have made the caveat explicit rather than implying a guarantee that is not there: a note on the `PostgreSQLSessionLockService` javadoc and on the `liquibase.useSessionLock` description recommending a dedicated non-pooled connection for Liquibase, and the failed-release `LockException` now says the lock stays held until the session ends instead of surfacing a bare SQL error.

On the stale lock you saw on a long-lived pod: if that was with the table-based `StandardLockService`, it is a different failure, a committed `LOCKED` row that nothing clears, which is what this PR is trying to remove. If it was the 4.x sessionlock extension on a pooled connection, it is likely the case above. Happy to dig further if you can say which one it was.

**Comment by ymajoros** [comment]
@coderabbitai review

**Comment by ymajoros** [comment]
@coderabbitai good catch, fixed in 10bd001a.

Every case passed a null catalog, so the branch folding the catalog into the advisory key was never exercised. That one matters more than a coverage gap: changing how catalog and schema are combined shifts the key, so an old and a new Liquibase would take two different advisory locks against the same lock table and quietly stop excluding each other. Pinned it on `appdb.migrations`, and pulled the shared try-lock stubbing into a helper so the two key tests do not duplicate it.

**Comment by filipelautert** [comment]
CI is unblocked now and it caught one thing: LiquibaseCommandLineTest "help output" needs the new global option added to its expected text. Could you update it?

File: liquibase-standard/src/test/java/liquibase/lockservice/PostgreSQLSessionLockServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        service.setChangeLogLockRecheckTime(7);
+        assertThat(service.getChangeLogLockWaitTimeMinutes()).isEqualTo(42);
+        assertThat(service.getChangeLogLockRecheckTimeSeconds()).isEqualTo(7);
+    }
+
+    /** Stubs pg_try_advisory_lock to grant the lock, and returns the statement to verify keys on. */
+    private static PreparedStatement stubSuccessfulTryLock(Connection connection) throws Exception {
+        PreparedStatement tryLock = mock(PreparedStatement.class);
+        ResultSet resultSet = mock(ResultSet.class);
+        when(connection.prepareStatement(TRY_LOCK_SQL)).thenReturn(tryLock);
```
[the remark this answers is not in the document]
Comment (ymajoros) [reply liquibase-standard/src/test/java/liquibase/lockservice/PostgreSQLSessionLockServiceTest.java:236]: False positive. `connection` here is `mock(Connection.class)`, so `prepareStatement` allocates no JDBC resource, and the `PreparedStatement` it returns is itself a mock used only to verify the advisory-lock key arguments. Closing them would mean closing mocks.

The same call appears inline in five other tests in this file and is not flagged there; only the two helper methods are, because the statement escapes the method and the "closed on method exit" check cannot see past that.

File: liquibase-standard/src/test/java/liquibase/lockservice/PostgreSQLSessionLockServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        when(resultSet.next()).thenReturn(true);
+        when(resultSet.getObject(1)).thenReturn(Boolean.TRUE);
+        return tryLock;
+    }
+
+    /** Stubs the pg_locks lookup used by listLocks(), optionally reporting one holder. */
+    private static void stubLockInfo(Connection connection, boolean locked) throws Exception {
+        PreparedStatement lockInfo = mock(PreparedStatement.class);
+        ResultSet resultSet = mock(ResultSet.class);
+        when(connection.prepareStatement(Mockito.startsWith("SELECT l.pid"))).thenReturn(lockInfo);
```
[the remark this answers is not in the document]
Comment (ymajoros) [reply liquibase-standard/src/test/java/liquibase/lockservice/PostgreSQLSessionLockServiceTest.java:247]: False positive. `connection` here is `mock(Connection.class)`, so `prepareStatement` allocates no JDBC resource, and the `PreparedStatement` it returns is itself a mock used only to verify the advisory-lock key arguments. Closing them would mean closing mocks.

The same call appears inline in five other tests in this file and is not flagged there; only the two helper methods are, because the statement escapes the method and the "closed on method exit" check cannot see past that.

**Comment by ymajoros** [comment]
@filipelautert done in e913203a, sorry for the slow turnaround.

`LiquibaseCommandLineTest` pins the whole `liquibase --help` text, so the new `--use-session-lock` block had to be added to it. I took picocli's actual rendering from the failure rather than hand-wrapping it, and it sits between `--use-procedure-schema` and `-v, --version` where the alphabetical ordering puts it.

Also rebased onto current main (it was 50 behind) and re-ran after the rebase, in case anything on main had shifted that same help text. It had not. Green locally on JDK 21: `liquibase-standard` 6510 tests and `liquibase-cli` 119, no failures.

One thing needing your hand: CodeQL posted two "Potential database resource leak" alerts on `PostgreSQLSessionLockServiceTest` (2328 and 2329). Both are false positives. `connection` there is `mock(Connection.class)`, so `prepareStatement` allocates no JDBC resource, and the `PreparedStatement` it returns is itself a mock used only to verify the advisory-lock key arguments. The same call appears inline in five other tests in that file and is not flagged; only the two helper methods are, because the statement escapes the method and the "closed on method exit" check cannot see past it. I have replied on both threads and resolved them, but dismissing the alerts in the Security tab needs someone with access to it.

Still open from the earlier round, both your call rather than mine:

1. `supportsAdvisoryLocks()` defaults to true on `AbstractPostgresDatabase` as you suggested, which is the opposite of the four snapshot-ability hooks next to it that default false and have `PostgresDatabase` opt up. Happy to flip it for uniformity.
2. A lost session lock now throws from `releaseLock()`, but that does not actually fail an `update`, because `AbstractUpdateCommandStep`, `AbstractFutureRollbackCommandStep` and `LockServiceCommandStep` all catch `LockException` and log it severe. Making it genuinely fail the run means touching that shared catch, which changes `StandardLockService` behaviour for everyone, so I kept it out of an opt-in PR. Follow-up if you want it.

**Comment by ymajoros** [comment]
@filipelautert both of my open questions are now decided, so there is nothing left to weigh up on this one.

**1. `supportsAdvisoryLocks()` now defaults to false** (ea641db5b), with `PostgresDatabase` opting up, matching the four snapshot-ability hooks beside it on `AbstractPostgresDatabase`. Uniformity with the file beats my "advisory locks are core since 8.2" argument, and it costs nothing: `PostgresDatabase` is the only direct subclass of the abstract, so every concrete database reports exactly what it did before. `PostgresDatabase` and the `EnterpriseDBDatabase` it parents say true, CockroachDB still overrides to false. I added an `EnterpriseDBDatabase` assertion to the test, because with the flip it is `PostgresDatabase` rather than the abstract that turns the capability on, and that is worth pinning.

**2. The "should a lost lock fail the run" question is out of this PR** and filed as #8008. It needs a change to the shared `LockException` catch in `AbstractUpdateCommandStep`, `AbstractFutureRollbackCommandStep` and `LockServiceCommandStep`, which would change `StandardLockService` behaviour for every existing user, so it is not something an opt-in PR should decide. The issue lays out the three options. Happy to implement whichever you prefer, separately.

**One thing I cannot do from here: CI needs your click.** Every workflow run on this branch sits at `action_required`, including the two pushes today. Approving a fork PR's run needs admin on the repo, and I only have read, so the runs will keep queuing unapproved no matter how many times I push. Whenever you get a moment, "Approve and run" on the latest head (ea641db5b) would confirm it.

Meanwhile, verified locally on JDK 21 against current main: `liquibase-standard` 6510 tests and `liquibase-cli` 119, no failures. That includes the `LiquibaseCommandLineTest` help output you flagged, which was the only real failure in the Sept 1 run.

**Comment by ymajoros** [comment]
@filipelautert short version of the comment above, since the important bit was buried in it: **this PR cannot go green on its own.**

Every workflow run on the branch sits at `action_required`, so `PR Gate` and `PR Labels`, two of the three required checks, never report at all. I only have read access, so the approve endpoint gives me a 403. One "Approve and run" on `ea641db5b` is all it needs, and then it is only the review.

Everything else is closed out: the help-output fix you asked for on 1 September, both of my open questions decided rather than left hanging, no unresolved threads. Rebased on current main and green locally on JDK 21, 6510 tests in `liquibase-standard` and 119 in `liquibase-cli`.

@wwillard7800 tagging you too in case you get to it first.
